cow_py/contracts/sign.py

```diff
@@ -0,0 +1,104 @@
+from enum import IntEnum
+from typing import List, NamedTuple, Union
+
+from eth_account import Account
+from eth_account.datastructures import SignedMessage
+from eth_account.signers.local import LocalAccount
+from eth_utils.conversions import to_hex
+from eth_utils.crypto import keccak
+from web3 import Web3
+
+from cow_py.contracts.order import (
+    CANCELLATIONS_TYPE_FIELDS,
+    ORDER_TYPE_FIELDS,
+    Order,
+    hash_typed_data,
+    normalize_order,
+)
+
+EIP1271_MAGICVALUE = to_hex(keccak(text="isValidSignature(bytes32,bytes)"))[:10]
+PRE_SIGNED = to_hex(keccak(text="GPv2Signing.Scheme.PreSign"))
+
+
+class SigningScheme(IntEnum):
+    # The EIP-712 typed data signing scheme. This is the preferred scheme as it
+    # provides more infomation to wallets performing the signature on the data
+    # being signed.
+    #
+    # https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-domainseparator
+    EIP712 = 0b00
+    # Message signed using eth_sign RPC call.
+    ETHSIGN = 0b01
+    # Smart contract signatures as defined in EIP-1271.
+    EIP1271 = 0b10
+    # Pre-signed order.
+    PRESIGN = 0b11
+
+
+class EcdsaSignature(NamedTuple):
+    scheme: SigningScheme
+    data: str
+
+
+class Eip1271SignatureData(NamedTuple):
+    verifier: str
+    signature: bytes
+
+
+class Eip1271Signature(NamedTuple):
+    scheme: SigningScheme
+    data: Eip1271SignatureData
+
+
+class PreSignSignature(NamedTuple):
+    scheme: SigningScheme
+    data: str
+
+
+Signature = Union[EcdsaSignature, Eip1271Signature, PreSignSignature]
+
+
+def ecdsa_sign_typed_data(
+    scheme, owner: LocalAccount, domain_data, message_types, message_data
+) -> SignedMessage:
+    return Account._sign_hash(
+        hash_typed_data(domain_data, message_types, message_data), owner.key
+    )
+
+
+def sign_order(
+    domain, order: Order, owner: LocalAccount, scheme: SigningScheme
+) -> EcdsaSignature:
+    normalized_order = normalize_order(order)
+    signed_data = ecdsa_sign_typed_data(
+        scheme, owner, domain, {"Order": ORDER_TYPE_FIELDS}, normalized_order
+    )
+    return EcdsaSignature(
+        scheme=scheme,
+        data=signed_data.signature.hex(),
+    )
+
+
+def sign_order_cancellation(domain, order_uid: Union[str, bytes], owner, scheme):
+    return sign_order_cancellations(domain, [order_uid], owner, scheme)
```

Comment:
There is no type for domain, owner and scheme - I expected pyright to catch this?